From 63a893606bb065b172dbada79a6be9c76bc3fe20 Mon Sep 17 00:00:00 2001
From: Chris Harrelson <chrishtr@chromium.org>
Date: Thu, 21 Dec 2023 22:13:13 +0000
Subject: [PATCH] Don't invalidate a11y for c-v subtrees just because of style recalc.

Instead, invalidate only when they are locked or unlocked.

Bug: 1512170

Change-Id: Ie4c433b8c1bad798ca3231abfaa8ff5c23c3d08d
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5137382
Commit-Queue: Chris Harrelson <chrishtr@chromium.org>
Reviewed-by: Vladimir Levin <vmpstr@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1240343}
---

--- a/third_party/blink/renderer/core/display_lock/display_lock_context.cc
+++ b/third_party/blink/renderer/core/display_lock/display_lock_context.cc
@@ -580,8 +580,9 @@
 
   // We also need to notify the AX cache (if it exists) to update the childrens
   // of |element_| in the AX cache.
-  if (AXObjectCache* cache = element_->GetDocument().ExistingAXObjectCache())
-    cache->ChildrenChanged(element_);
+  if (auto* ax_cache = element_->GetDocument().ExistingAXObjectCache()) {
+    ax_cache->SubtreeIsAttached(element_);
+  }
 
   // Schedule ContentVisibilityAutoStateChange event if needed.
   ScheduleStateChangeEventIfNeeded();
--- a/third_party/blink/renderer/core/dom/element.cc
+++ b/third_party/blink/renderer/core/dom/element.cc
@@ -3653,7 +3653,7 @@
       new_style = nullptr;
       NotifyAXOfAttachedSubtree();
     } else {
-      if (!new_style->IsContentVisibilityVisible()) {
+      if (!old_style && !new_style->IsContentVisibilityVisible()) {
         NotifyAXOfAttachedSubtree();
       }
       if (new_style->IsContainerForSizeContainerQueries()) {
